src/io/metadata.cpp

```diff
@@ -528,6 +557,33 @@ void Metadata::LoadWeights() {
   weight_load_from_file_ = true;
 }
 
+void Metadata::LoadPositions() {
+  num_positions_ = 0;
+  std::string position_filename(data_filename_);
+  // default position file name
+  position_filename.append(".position");
```

Comment:
Instead of hard-coding this assumption deep inside this method, I think we should expose configuration like `position_bias_factor_file`, similar to how you can provide `train_data_file` with a string pointing to a file containing the training data: https://lightgbm.readthedocs.io/en/latest/Parameters.html#data.

That way:

* it would be possible to load positions from a file even if the training data is provided via an in-memory `Dataset`
* users could name the file whatever they want

Would you please consider that?

Reply:
I agree. To support something like `positions=<numpy.array>` in Python API would require change of the C API, which can be a breaking change. However, if we allow the `position_filename` to be specified by users, the user can specify different position files for the same dataset. Which is somewhat similar with that `weights` can be specified differently for the same dataset with the parameter `weight` in `lgb.Dataset`.

I can help with that.

Reply:
> To support something like positions=<numpy.array> in Python API would require change of the C API, which can be a breaking change

I don't agree with that. Couldn't the existing methods in the C API for initializing `Dataset` just not have `positions`, and then we could require that `positions` be set after construction with `LGBM_DatasetSetField()`?

https://github.com/microsoft/LightGBM/blob/714039681a9502ce6644357d38ec9143bf5a7852/include/LightGBM/c_api.h#L534-L538

And on the Python side, we could make `positions` an optional argument everywhere that `init_score` and `weights` are provided, accepting exactly the same data types, and then call `LGBM_DatasetSetField()` to set `positions` on the `Dataset` on the C/C++ side.

That's exactly how `weight` is set on the Dataset from the Python side.

https://github.com/microsoft/LightGBM/blob/714039681a9502ce6644357d38ec9143bf5a7852/python-package/lightgbm/basic.py#L1879-L1880

https://github.com/microsoft/LightGBM/blob/714039681a9502ce6644357d38ec9143bf5a7852/python-package/lightgbm/basic.py#L2631-L2633

https://github.com/microsoft/LightGBM/blob/714039681a9502ce6644357d38ec9143bf5a7852/python-package/lightgbm/basic.py#L2404-L2409

Reply:
@jameslamb I see. What do you think if we first expose the `position_bias_factor_file` in this PR, and leave the change in the python API in another PR? Allowing the `position` to be used in the same way as `group` or `init_score` requires a lot of changes. For example, we may even want to allow the `position` to be specified in one column of a LibSVM file. These requires many changes to the Python API and dataloader.

Reply:
That's file as long as all that follow-up work is documented in issues in this project's backlog, and as long as you and @metpavel are committing to working on them in the near future.

Otherwise, I'm afraid we'll be taking on this code path into the project that users will not really be able to take advantage of expect under a very specific set of circumstances.

Reply:
Indeed, the valid set does not require the position file for now. But there could be some cases where a position file could be used in the future. For example, we add some metrics to evaluate the position unbiased ranking loss.

Also, even though the valid set does not require the position file, I think keep it consistent with  weight, query, initial score should be a good practice.

Reply:
> add some metrics to evaluate the position unbiased ranking loss

Ah ok, makes sense to me. Thank you!

> even though the valid set does not require the position file, I think keep it consistent with weight, query, initial score should be a good practice.

I strongly agree! It was my mistake that I didn't realize we had similar conventions for file names for those other metadata types, sorry about that.